cartographer_ros/cartographer_ros/assets_writer_main.cc

```diff
@@ -205,25 +168,51 @@ void Run(const string& pose_graph_filename,
       const double duration_in_seconds =
           (view.getEndTime() - begin_time).toSec();
 
+      // We make sure that tf_messages are published before any data messages,
+      // so that tf lookups always work and that tf_buffer has a small cache
+      // size - because it gets very inefficient with a large one.
+      std::deque<rosbag::MessageInstance> delayed_messages;
       for (const rosbag::MessageInstance& message : view) {
-        std::unique_ptr<carto::io::PointsBatch> points_batch;
-        if (message.isType<sensor_msgs::PointCloud2>()) {
-          points_batch = HandleMessage(
-              *message.instantiate<sensor_msgs::PointCloud2>(), tracking_frame,
-              tf_buffer, transform_interpolation_buffer);
-        } else if (message.isType<sensor_msgs::MultiEchoLaserScan>()) {
-          points_batch = HandleMessage(
-              *message.instantiate<sensor_msgs::MultiEchoLaserScan>(),
-              tracking_frame, tf_buffer, transform_interpolation_buffer);
-        } else if (message.isType<sensor_msgs::LaserScan>()) {
-          points_batch = HandleMessage(
-              *message.instantiate<sensor_msgs::LaserScan>(), tracking_frame,
-              tf_buffer, transform_interpolation_buffer);
+        if (FLAGS_use_bag_transforms && message.isType<tf2_msgs::TFMessage>()) {
+          auto tf_message = message.instantiate<tf2_msgs::TFMessage>();
+          for (const auto& transform : tf_message->transforms) {
+            try {
+              tf_buffer.setTransform(transform, "unused_authority",
+                                     message.getTopic() == kTfStaticTopic);
+            } catch (const tf2::TransformException& ex) {
+              LOG(WARNING) << ex.what();
+            }
+          }
         }
-        if (points_batch != nullptr) {
-          points_batch->trajectory_id = trajectory_id;
-          pipeline.back()->Process(std::move(points_batch));
+
+        while (!delayed_messages.empty() &&
+               delayed_messages.front().getTime() <
+                   message.getTime() - ::ros::Duration(1.)) {
```

Comment:
Is there any significance to this constant? Can you explain why it needs to be 1?

Reply:
I was confused about the same thing some time ago. Have a look here:
https://github.com/googlecartographer/cartographer_ros/commit/8b239ded863d4f3fcf4134ce2b575e7f8711df44#commitcomment-22207271